rfcs/006-package-management.md

```diff
@@ -0,0 +1,243 @@
+---
+feature: package management
+start-date: 2024-06-28
+author: Joe Neeman
+---
+
+# Package management
+
+People want to reuse code, but Nickel doesn't currently have a good way to do
+it. We should have a way to fetch packages and make them easily available to
+Nickel code. The mechanism needs to be predictable (it should fetch the code
+that the user expects to fetch) and reliable (if it works on my machine it
+should work on your machine).
+
+## The manifest file
+
+We will require a manifest file in order to import packages. Manifest files must
+be named `package.ncl`, and they are found by searching up from the file being
+evaluated. That is, when the user invokes `nickel export path/to/foo.ncl`, we
+look for a manifest at `path/to/package.ncl` and then at `path/package.ncl`, and
+so on.
+
+The manifest file format is defined by the contract `std.package.Manifest`,
+while is defined as
+
+```nickel
+{
+  name | String,
+  version | Semver,
+  nickel-version | Semver,
+
+  dependencies
+    | { _:
+      [|
+        'Path String,
+        'Git { url | String, branch | optional | String, rev | optional | String },
+        'Index { name | String, version | Semver },
```

Comment:
Ok, this makes it harder to have a .toml file. Another solution is to emit a warning when the file isn't a full normal form, and ask the user if they want to continue. Or something like that. On the other hand `flake.nix` can have pretty complex logic, so maybe it's ok.

Reply:
Yeah, the ability to have a nice contract was my main motivation for proposing a nickel manifest. There are definitely some advantages to a plain-data format, though...